conda_libmamba_solver/solver.py

```diff
@@ -475,21 +478,22 @@ def _specs_to_tasks_add(self, in_state: SolverInputState, out_state: SolverOutpu
                 tasks[("USERINSTALLED", api.SOLVER_USERINSTALLED)].append(installed_spec_str)
 
             # These specs are explicit in some sort of way
-            if pinned:
+            if pinned and not pinned.is_name_only_spec:
                 # these are the EXPLICIT pins; conda also uses implicit pinning to
                 # constrain updates too but those can be overridden in case of conflicts.
-                if pinned.is_name_only_spec:
-                    # pins need to constrain in some way, otherwide is undefined behaviour
-                    pass
-                elif requested and not requested.match(pinned):
-                    # We don't pin; requested and pinned are different and incompatible,
-                    # requested wins and we let that happen in the next block
-                    pass
-                else:
-                    tasks[("ADD_PIN", api.SOLVER_NOOP)].append(self._spec_to_str(pinned))
-
+                # name-only pins are treated as locks when installed, see below
+                tasks[("ADD_PIN", api.SOLVER_NOOP)].append(self._spec_to_str(pinned))
+            # in libmamba, pins and installs are compatible tasks (pin only constrains,
+            # does not 'request' a package). In classic, pins were actually targeted installs
+            # so they were exclusive
             if requested:
-                spec_str = self._spec_to_str(requested)
+                if requested.is_name_only_spec and pinned and not pinned.is_name_only_spec:
+                    # for name-only specs, this is a no-op; we already added the pin above
+                    # but we will constrain it again in the install task to have better
+                    # error messages if not solvable
+                    spec_str = self._spec_to_str(pinned)
```

Comment:
Note this is purely for UX purposes. The solver doesn't care if we do either of:

A:

```
install("python")
pin("python=3.7")
```

B:

```
install("python=3.7")
pin("python=3.7")
```

However, B will give (as of `libmamba 1.5.x`) a clearer error message if the request happens to be unsolvable.